<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkST0151ENSkillsNetwork20531532-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo"  />
    </a>
</p>


# **Regression Analysis**


Estimated time needed: **30** minutes


The goal of regression analysis is to describe the relationship between one set of variables called the dependent variables, and another set of variables, called independent or explanatory variables. When there is only one explanatory variable, it is called simple regression.


## Objectives


After completing this lab you will be able to:


*   Import Libraries
*   Regression analysis in place of the t-test
*   Regression analysis in place of ANOVA
*   Regression analysis in place of correlation


***


## Import Libraries


All Libraries required for this lab are listed below. The libraries pre-installed on Skills Network Labs are commented. If you run this notebook in a different environment, e.g. your desktop, you may need to uncomment and install certain libraries.


In [1]:
#install specific version of libraries used in lab
#! mamba install pandas==1.3.3
#! mamba install numpy=1.21.2
#! mamba install scipy=1.7.1-y
#!  mamba install seaborn=0.9.0-y
#!  mamba install matplotlib=3.4.3-y
#!  mamba install statsmodels=0.12.0-y

Import the libraries we need for the lab


In [2]:
import piplite
await piplite.install(['numpy'],['pandas'])

In [3]:
import numpy as np
import pandas as pd
import statsmodels.api as sm

In [4]:
from js import fetch
import io

URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-ST0151EN-SkillsNetwork/labs/teachingratings.csv'
resp = await fetch(URL)
ratings_url = io.BytesIO((await resp.arrayBuffer()).to_py())

Read in the csv file from the URL using the request library


In [5]:
ratings_df = pd.read_csv(ratings_url)
ratings_df

minority  age  gender credits    beauty  eval division native tenure  \
0        yes   36  female    more  0.289916   4.3    upper    yes    yes   
1        yes   36  female    more  0.289916   3.7    upper    yes    yes   
2        yes   36  female    more  0.289916   3.6    upper    yes    yes   
3        yes   36  female    more  0.289916   4.4    upper    yes    yes   
4         no   59    male    more -0.737732   4.5    upper    yes    yes   
..       ...  ...     ...     ...       ...   ...      ...    ...    ...   
458       no   37    male    more  0.933396   3.5    upper    yes    yes   
459       no   58  female    more  0.111563   3.5    upper    yes    yes   
460       no   42    male    more -0.900580   4.0    upper    yes    yes   
461       no   39    male    more  0.643014   4.3    upper    yes    yes   
462       no   51  female    more  0.391822   3.0    upper    yes    yes   

     students  allstudents  prof  PrimaryLast  vismin  female  single_credit  \
0          24           43     1            0       1       1              0   
1          86          125     1            0       1       1              0   
2          76          125     1            0       1       1              0   
3          77          123     1            1       1       1              0   
4          17           20     2            0       0       0              0   
..        ...          ...   ...          ...     ...     ...            ...   
458        15           17    40            1       0       0              0   
459        26           34    47            1       0       1              0   
460        45           86    61            1       0       0              0   
461        22           29    62            1       0       0              0   
462        47           67    69            1       0       1              0   

     upper_division  English_speaker  tenured_prof  
0                 1                1             1  
1                 1                1             1  
2                 1                1             1  
3                 1                1             1  
4                 1                1             1  
..              ...              ...           ...  
458               1                1             1  
459               1                1             1  
460               1                1             1  
461               1                1             1  
462               1                1             1  

[463 rows x 19 columns]

## Lab Exercises


In this section, you will learn how to run regression analysis in place of the t-test, ANOVA, and correlation


### Regression with T-test: Using the teachers rating data set, does gender affect teaching evaluation rates?


Initially, we had used the t-test to test if there was a statistical difference in evaluations for males and females, we are now going to use regression. We will state the null hypothesis:

*   $H\_0: β1$ = 0 (Gender has no effect on teaching evaluation scores)
*   $H\_1: β1$ is not equal to 0 (Gender has an effect on teaching evaluation scores)


We will use the female variable. female = 1 and male = 0


In [6]:
## X is the input variables (or independent variables)
X = ratings_df['female']
## y is the target/dependent variable
y = ratings_df['eval']
## add an intercept (beta_0) to our model
X = sm.add_constant(X) 

model = sm.OLS(y, X).fit()
predictions = model.predict(X)

# Print out the statistics
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   eval   R-squared:                       0.022
Model:                            OLS   Adj. R-squared:                  0.020
Method:                 Least Squares   F-statistic:                     10.56
Date:                Mon, 11 Sep 2023   Prob (F-statistic):            0.00124
Time:                        18:17:17   Log-Likelihood:                -378.50
No. Observations:                 463   AIC:                             761.0
Df Residuals:                     461   BIC:                             769.3
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          4.0690      0.034    121.288      0.000       4.003       4.135
female        -0.1680      0.052     -3.250      0.001      -0.270      -0.066
==============================================================================
Omnibus:                       17.625   Durbin-Watson:                   1.209
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               18.970
Skew:                          -0.496   Prob(JB):                     7.60e-05
Kurtosis:                       2.981   Cond. No.                         2.47
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Conclusion:** Like the t-test, the p-value is less than the alpha (α) level = 0.05, so we reject the null hypothesis as there is evidence that there is a difference in mean evaluation scores based on gender. The coefficient -0.1680 means that females get 0.168 scores less than men.


### Regression with ANOVA: Using the teachers' rating data set, does beauty  score for instructors  differ by age?


State the Hypothesis:

*   $H\_0: µ1 = µ2 = µ3$ (the three population means are equal)
*   $H\_1:$ At least one of the means differ


Then we group the data like we did with ANOVA


In [7]:
ratings_df.loc[(ratings_df['age'] <= 40), 'age_group'] = '40 years and younger'
ratings_df.loc[(ratings_df['age'] > 40)&(ratings_df['age'] < 57), 'age_group'] = 'between 40 and 57 years'
ratings_df.loc[(ratings_df['age'] >= 57), 'age_group'] = '57 years and older'
ratings_df

minority  age  gender credits    beauty  eval division native tenure  \
0        yes   36  female    more  0.289916   4.3    upper    yes    yes   
1        yes   36  female    more  0.289916   3.7    upper    yes    yes   
2        yes   36  female    more  0.289916   3.6    upper    yes    yes   
3        yes   36  female    more  0.289916   4.4    upper    yes    yes   
4         no   59    male    more -0.737732   4.5    upper    yes    yes   
..       ...  ...     ...     ...       ...   ...      ...    ...    ...   
458       no   37    male    more  0.933396   3.5    upper    yes    yes   
459       no   58  female    more  0.111563   3.5    upper    yes    yes   
460       no   42    male    more -0.900580   4.0    upper    yes    yes   
461       no   39    male    more  0.643014   4.3    upper    yes    yes   
462       no   51  female    more  0.391822   3.0    upper    yes    yes   

     students  allstudents  prof  PrimaryLast  vismin  female  single_credit  \
0          24           43     1            0       1       1              0   
1          86          125     1            0       1       1              0   
2          76          125     1            0       1       1              0   
3          77          123     1            1       1       1              0   
4          17           20     2            0       0       0              0   
..        ...          ...   ...          ...     ...     ...            ...   
458        15           17    40            1       0       0              0   
459        26           34    47            1       0       1              0   
460        45           86    61            1       0       0              0   
461        22           29    62            1       0       0              0   
462        47           67    69            1       0       1              0   

     upper_division  English_speaker  tenured_prof                age_group  
0                 1                1             1     40 years and younger  
1                 1                1             1     40 years and younger  
2                 1                1             1     40 years and younger  
3                 1                1             1     40 years and younger  
4                 1                1             1       57 years and older  
..              ...              ...           ...                      ...  
458               1                1             1     40 years and younger  
459               1                1             1       57 years and older  
460               1                1             1  between 40 and 57 years  
461               1                1             1     40 years and younger  
462               1                1             1  between 40 and 57 years  

[463 rows x 20 columns]

Use OLS function from the statsmodel library


In [8]:
from statsmodels.formula.api import ols
lm = ols('beauty ~ age_group', data = ratings_df).fit()
table= sm.stats.anova_lm(lm)
print(table)

              df      sum_sq    mean_sq          F        PR(>F)
age_group    2.0   20.422744  10.211372  17.597559  4.322549e-08
Residual   460.0  266.925153   0.580272        NaN           NaN


**Conclusion:** We can also see the same values for ANOVA like before and we will reject the null hypothesis since the p-value is less than 0.05 there is significant evidence that at least one of the means differ.


### Regression with ANOVA option 2


Create dummy variables - A dummy variable is a numeric variable that represents categorical data, such as gender, race, etc. Dummy variables are dichotomous, i.e they can take on only two quantitative values.


In [12]:
X = pd.get_dummies(ratings_df[['age_group']])
X

age_group_40 years and younger  age_group_57 years and older  \
0                                 1                             0   
1                                 1                             0   
2                                 1                             0   
3                                 1                             0   
4                                 0                             1   
..                              ...                           ...   
458                               1                             0   
459                               0                             1   
460                               0                             0   
461                               1                             0   
462                               0                             0   

     age_group_between 40 and 57 years  
0                                    0  
1                                    0  
2                                    0  
3                                    0  
4                                    0  
..                                 ...  
458                                  0  
459                                  0  
460                                  1  
461                                  0  
462                                  1  

[463 rows x 3 columns]

In [13]:
y = ratings_df['beauty']
## add an intercept (beta_0) to our model
X = sm.add_constant(X) 

model = sm.OLS(y, X).fit()
predictions = model.predict(X)

# Print out the statistics
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 beauty   R-squared:                       0.071
Model:                            OLS   Adj. R-squared:                  0.065
Method:                 Least Squares   F-statistic:                     11.71
Date:                Mon, 11 Sep 2023   Prob (F-statistic):           2.10e-07
Time:                        18:31:25   Log-Likelihood:                -529.47
No. Observations:                 463   AIC:                             1067.
Df Residuals:                     459   BIC:                             1083.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================================
                                        coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------
const                                 0.0138      0.028      0.495      0.621      -0.041       0.069
age_group_40 years and younger        0.3224      0.058      5.568      0.000       0.209       0.436
age_group_57 years and older         -0.2596      0.056     -4.616      0.000      -0.370      -0.149
age_group_between 40 and 57 years    -0.0489      0.045     -1.080      0.281      -0.138       0.040
==============================================================================
Omnibus:                       11.586   Durbin-Watson:                   0.434
Prob(Omnibus):                  0.003   Jarque-Bera (JB):               12.114
Skew:                           0.394   Prob(JB):                      0.00234
Kurtosis:                       2.913   Cond. No.                     1.12e+15
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 5.11e-28. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

You will get the same results and conclusion


### Correlation: Using the teachers' rating dataset, Is teaching evaluation score correlated with beauty score?


In [14]:
## X is the input variables (or independent variables)
X = ratings_df['beauty']
## y is the target/dependent variable
y = ratings_df['eval']
## add an intercept (beta_0) to our model
X = sm.add_constant(X) 

model = sm.OLS(y, X).fit()
predictions = model.predict(X)

# Print out the statistics
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   eval   R-squared:                       0.036
Model:                            OLS   Adj. R-squared:                  0.034
Method:                 Least Squares   F-statistic:                     17.08
Date:                Mon, 11 Sep 2023   Prob (F-statistic):           4.25e-05
Time:                        18:33:41   Log-Likelihood:                -375.32
No. Observations:                 463   AIC:                             754.6
Df Residuals:                     461   BIC:                             762.9
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          3.9983      0.025    157.727      0.000       3.948       4.048
beauty         0.1330      0.032      4.133      0.000       0.070       0.196
==============================================================================
Omnibus:                       15.399   Durbin-Watson:                   1.238
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               16.405
Skew:                          -0.453   Prob(JB):                     0.000274
Kurtosis:                       2.831   Cond. No.                         1.27
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Conclusion:** p < 0.05 there is evidence of correlation between beauty and evaluation scores


## Practice Questions


### Question 1: Using the teachers' rating data set, does tenure affect beauty scores?

*   Use α = 0.05


In [20]:
### insert code here
import statsmodels.api as sm

x = ratings_df['tenured_prof'] #要抓(1/0)的資料 不是(yes/no) #the OLS library doesn't recognize texts
y = ratings_df['beauty']

x = sm.add_constant(x)

model = sm.OLS(y, x).fit()
predictions = model.predict(x)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 beauty   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                    0.1689
Date:                Mon, 11 Sep 2023   Prob (F-statistic):              0.681
Time:                        18:43:54   Log-Likelihood:                -546.45
No. Observations:                 463   AIC:                             1097.
Df Residuals:                     461   BIC:                             1105.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const            0.0284      0.078      0.363      0.717      -0.125       0.182
tenured_prof    -0.0364      0.089     -0.411      0.681      -0.210       0.138
==============================================================================
Omnibus:                       23.184   Durbin-Watson:                   0.461
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               23.229
Skew:                           0.507   Prob(JB):                     9.03e-06
Kurtosis:                       2.583   Cond. No.                         4.05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Double-click **here** for a hint.

<!-- The hint is below:
## State Hypothesis
Null Hypothesis: Mean beauty scores for tenured and non-tenured instructors are equal
Alternative Hypothesis: There is a difference in mean beauty scores for tenured and non-tenured instructors

## use the dummy variable for tenure - the OLS library doesn't recognize texts
X = ratings_df['tenured_prof']
-->


Double-click **here** for the solution.

<!-- The answer is below:
## put beauty scores in a list
y = ratings_df['beauty']
## add an intercept (beta_0) to our model
X = sm.add_constant(X) 

model = sm.OLS(y, X).fit()
predictions = model.predict(X)

# Print out the statistics
model.summary()
p-value is greater than 0.05, so we fail to reject the null hypothesis as there is no evidence that the mean difference of tenured and untenured instructors are different
-->


### Question 2: Using the teachers' rating data set, does being an English speaker affect the number of students assigned to professors?

*   Use "allstudents"
*   Use α = 0.05 and α = 0.1


In [23]:
## insert code here
x=ratings_df['English_speaker']
y=ratings_df['allstudents']

x = sm.add_constant(x)

model = sm.OLS(y, x).fit()
predictions = model.predict(x)
print(model.summary())

print('At α = 0.05, we do not reject the H0, which indicates that being an native speaker do not affect the number of students assigned to professors')
print('At α = 0.1, we reject the H0, which indicates that being an native speaker does affect the number of students assigned to professors')

                            OLS Regression Results                            
Dep. Variable:            allstudents   R-squared:                       0.007
Model:                            OLS   Adj. R-squared:                  0.005
Method:                 Least Squares   F-statistic:                     3.476
Date:                Mon, 11 Sep 2023   Prob (F-statistic):             0.0629
Time:                        18:55:29   Log-Likelihood:                -2654.2
No. Observations:                 463   AIC:                             5312.
Df Residuals:                     461   BIC:                             5321.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const              29.6071     14.150     

Double-click **here** for a hint.

<!-- The hint is below:
## State Hypothesis
Null Hypothesis: Mean number of students assigned to native English speakers vs non-native English speakers are equal
Alternative Hypothesis: There is a difference in mean number of students assigned to native English speakers vs non-native English speakers

## Is the instructor a native English speaker - make sure to use the binary variable "English speaker"
X = ratings_df['English_speaker']
## You can use the students or all students variable
y = ratings_df['allstudents']
-->


Double-click **here** for the solution.

<!-- The answer is below:
## add an intercept (beta_0) to our model
X = sm.add_constant(X) 

model = sm.OLS(y, X).fit()
predictions = model.predict(X)

# Print out the statistics
model.summary()

At α = 0.05, p-value is greater, we fail to reject the null hypothesis as there is no evidence that being a native English speaker or a non-native English speaker affects the number of students assigned to an instructor.
At α = 0.1, p-value is less, we reject the null hypothesis as there is evidence that there is a significant difference of mean number of students assigned to native English speakers vs non-native English speakers.
-->


### Question 3: Using the teachers' rating data set, what is the correlation between the number of students who participated in the evaluation survey and evaluation scores?

*   Use "students" variable


In [25]:
## insert code here
x=ratings_df['students']
y=ratings_df['eval']

x = sm.add_constant(x)

model = sm.OLS(y, x).fit()
predictions = model.predict(x)
print(model.summary())
print('Do not reject H0, indicating that there is no correlation between the number of students who participated in the evaluation survey and evaluation scores')

                            OLS Regression Results                            
Dep. Variable:                   eval   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                 -0.001
Method:                 Least Squares   F-statistic:                    0.5806
Date:                Mon, 11 Sep 2023   Prob (F-statistic):              0.446
Time:                        19:03:07   Log-Likelihood:                -383.46
No. Observations:                 463   AIC:                             770.9
Df Residuals:                     461   BIC:                             779.2
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          3.9823      0.033    119.689      0.0

Double-click **here** for a hint.

<!-- The hint is below:
## create a list of students and evaluation socres
X = ratings_df['students']
y = ratings_df['eval']
-->


Double-click **here** for the solution.

<!-- The rest of the answer is below:
## add an intercept (beta_0) to our model
X = sm.add_constant(X) 

model = sm.OLS(y, X).fit()
predictions = model.predict(X)

# Print out the statistics
model.summary()

R-square is 0.001, R will be √0.001, correlation coefficient is 0.03 (close to 0). There is a very weak correlation between the number of students who participated in the evaluation survey and evaluation scores
-->


## Authors

[Pratiksha Verma](https://www.linkedin.com/in/pratiksha-verma-6487561b1/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkST0151ENSkillsNetwork956-2022-01-01)

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By      | Change Description      |
| ----------------- | ------- | -------------   | ----------------------- |
| 2022-11-09        | 1.0     | Pratiksha Verma | Converted initial version to Jupyterlite|


<hr>

## <h3 align="center"> IBM Corporation 2022. All rights reserved. <h3/>
